src/hardware/mixer.cpp

```diff
@@ -1693,45 +1694,130 @@ AudioFrame MixerChannel::ApplyCrossfeed(const AudioFrame frame) const
 	return {a.left + b.left, a.right + b.right};
 }
 
-MixerChannel::Sleeper::Sleeper(MixerChannel& c) : channel(c) {}
+void MixerChannel::ConfigureFadeout(const std::string_view prefs)
+{
+	sleeper.ConfigureFadeout(prefs);
+}
 
-// Records if samples had a magnitude great than 1. This is a one-way street;
-// once "had_noise" is tripped, it can only be reset after waking up.
-void MixerChannel::Sleeper::Listen(const AudioFrame frame)
+void MixerChannel::Sleeper::ConfigureFadeout(const std::string_view prefs)
 {
-	if (had_noise) {
+	auto set_wait_and_fade = [&](const int wait_ms, const int fade_ms) {
+		fadeout_or_sleep_after_ms = check_cast<uint16_t>(wait_ms);
+
+		fadeout_decrement_per_ms = 1.0f / static_cast<uint16_t>(fade_ms);
```

Comment:
Successively decrementing can leave you with a close to zero but not quite zero number. I know there were problems with such denormal floats on earlier CPUs, so make sure to set the level explicitly to 0.0f once the fade is over. The sometime an extra MUL might be more performant than a SUB and an extra IF, but in this case we're doing so little work it doesn't matter either way...

Btw, I don't think a float SUB is really faster than a MUL in actual algos when pipelining is involved, etc. A MUL is 1 cycle usually.

Reply:
> Successively decrementing can leave you with a close to zero but not quite zero number. 

I think you mean successively multiplying (by a percent)?

This is a decrement that's subtracted each millisecond, and just keeps decrementing until it dips below 0 and is clamped to zero.

So for example, a 10ms fade out will produce a decrement of 0.1/ms.

The multiplier values each millisecond pass become:

1.0 - 0.1 * (1ms into fade) = 0.9
1.0 - 0.1 * (2ms into fade) = 0.8
...
1.0 - 0.1 * (9ms into fade) = 0.1
1.0 - 0.1 * (10ms into fade) = 0.0
...
next step goes negative, and is clamped to zero.

Reply:
Yeah, that's fine, maybe I was a bit overthinking it 😄 Glad to see the clamp to zero there too 😄